### ISSUE TITLE
Using Love2D as a Lua module from another Lua runtime.

### ISSUE THREAD

**Issue by noncom** [love2d/love#2103]
I'm trying to use Love2D as a Lua module from another Lua program, so that my runtime is not hosted by Love2D.
I've found 2 instructions where people tried that, and it worked for them:
- https://love2d.org/forums/viewtopic.php?t=86145
- https://gist.github.com/markandgo/dfa7d4c1fc7b81da2ed5

So I modified `package.cpath`  to include the love's dll location,
However when I'm trying this:
```lua
local love = require 'love'
```

I'm getting:
```
error loading module 'love' from file 'D:\work\_love2d\megasource\build\love\Release\love.dll':
        The specified module could not be found.
```
Why could that be the case?

Note: I've compiled Love2D from the source, with Lua 5.4.7, and my main program also runs on Lua 5.4.7.

**Comment by MikuAuahDark** [comment]
First thing I'd check is if the symbol `luaopen_love` is exported. You can check this using [Dependencies](https://github.com/lucasg/Dependencies). It's also worth noting that the architecture must match (e.g. your x64 app can only load x64 LOVE).

**Comment by noncom** [comment]
Thanks! So that's what I tried, and what I have atm:

#### DLL search
- Checked `love.dll` with the Dependencies app, and it does show `luaopen_love` and other `luaopen_*` functions
- The Dependencies app also shows some dlls that `love.dll` depends on:
  - In the same folder: `OpenAL32.dll`, `SDL2.dll`, `lua.dll` -- all of these are indeed in the same folder
  - A bunch of dlls in `C:/Windows/system32`
- I modified `package.cpath`:
  -  To include the folder with `love.dll` and friends. This makes Lua being able to find it at all, because otherwise the error is about it being unable to find the file at all. So that part works.
  - Also added the `system32` folder just to be sure.
 - The fact that `love.dll` has its own friend `lua.dll` is a bit suspicious:
   - I'm not sure if it's going to recognize my runtime and use it or try to load that copy of Lua? Or what should happen at all? Just in case if that other runtime is somehow used, and my `pacakge.cpath`  doesn't affect it, I also set `LUA_CPATH` env var to the same value.
   - Both my main runtime, and the love's friend `lua.dll` are of the same Lua 5.4.7 version.

#### Compilation
- I explicitly compiled everything for 64bit, so there should be no problem with this. Also verified by testing it [this way](https://superuser.com/a/889267).
- I compiled my main runtime with MinGW-w64. Love2D compiles with MSVC instead, but I don't think that this should be a problem. [This guide](https://nullprogram.com/blog/2021/05/31/) says that marking functions with `__declspec(dllexport)` should resolve any issues between MinGW-w64 and MSVC compiled binaries. This is exaclty what Love2D already does: [see-1](https://github.com/love2d/love/blob/11.5/src/modules/love/love.h#L37), [see-2](https://github.com/love2d/love/blob/11.5/src/common/config.h#L108).
- The main runtime is actually a 64bit build of Fennel, but I'm currently excluding this fact from factors since in essence Fennel is just a preloaded Lua script that doesn't change anything on the binary side. Also, trying a vanilla Lua build from https://joedf.github.io/LuaBuilds/ results in absolutely same behavior.
- I tried building and requiring [luv](https://github.com/luvit/luv) the same way as Love2D now, and it worked without issues. So that's a counter-example. It's too, however, compiled with MinGW-w64.

#### I'm not sure what else to try

The question is: Why can't Lua find the module inside the dll even if it has the `luaopen_*` functions exposed?
Currently I have no answer to that.

Maybe someone with more experience in all these topics would have a guess of what's going on.

Some of other ideas I had, which I don't think make much sense, but which I might try out in the absense of better ideas:
- Maybe I could try building the whole thing around LuaJIT, and although I need the latest Lua, maybe that would yield some insights
- Trying different toolchains, like compile the host Fennel or Lua with MSVC. Or try somehow building Love2D with MinGW-w64, but looking at all the CMakeLists I wouldn't want to try that.
- By default Fennel includes a very simple C launcher program to facilitate its standalone build. Maybe I could try building the Fennel executable with Love2D as the launcher. But this might need some additional work that I can't easily estimate. I'm not sure if it will actually help to find out the cause of the problem, and since I would really need Love2D to be an optionally enableable module, this is also a bit off-track, as I see it.

**Comment by learn-more** [comment]
If you suspect dll loading issues (and are on windows), you might try this:

1. Download [GFlagsX](https://github.com/zodiacon/GflagsX/releases)
2. Open GFlagsX, navigate to the 'Image' tab
3. Select your executable from the dropdown, or select it with the 'New Image' button.
4. Under 'User Mode' tick the 'Show Loader Snaps' option
5. Click 'Apply Flags'
6. Run your process under a debugger, and look for any dll loading failures.
7. Don't forget to disable the 'Show Loader Snaps' again, because it _will_ slow down your executable.

**Comment by slime73** [comment]
The following code works fine for me on Windows when running through LuaJIT's standalone interpreter:

```Lua
package.cpath = "C:\\Program Files\\LOVE\\?.dll;"..package.cpath

arg[0] = nil -- force nogame

local love = require("love")
local boot = require("love.boot")

local c = coroutine.create(boot)

while coroutine.status(c) ~= "dead" do
	coroutine.resume(c)
end

```

It sounds like you have a configuration issue with your non-standard build setup. The discord server or forums might be able to help, but this issue tracker is more for issues in love's own source code and build setup.

**Comment by noncom** [comment]
Thanks everyone for the suggestions, references, and help! I have learned some new and useful things from this.
I don't have a lot of experience with C/C++ and native stuff, so that was quite an adventure.

In the end it indeed turned out to be an issue with my setup, there were two problems:
- Fennel by default initially builds with a static Lua library, which later conflicts with what Love2D expects. This was something coming from how Fennel exe was being built.
- After I fixed that and compiled Fennel with Lua dynamically linked, another issue was discovered: a different version of Lua was leaking into the Love2D build and Fennel, but only during runtime. Thia happened because of how system variables got resolved by the running shells when running Fennel/Lua interpreter. During compilation the project was linking another version of Lua that was specified in the build files. And it of course still leaks from the PATH variable anytime unless I put all the binaries into the same folder, where the local files get priority over what's in the PATH.

That was quite a mess, I must admit. Just describing this for myself as a memo, and maybe if someone ever would have a similar kind of problem. Maybe I'll create an example repo with the working setup later when I finish setting up the rest of the project.
